torch/nn/quantization.py

```diff
@@ -0,0 +1,197 @@
+  scale = scale.detach()
+  # The scale will be zero if all values in a channel of x are zero. We use
+  # nan_to_num to ensure that x / scale * scale == x for this case.
+  x = torch.nan_to_num_(x / scale)
+  return x, scale
+
+
+def ste(forward):
+  """Returns 'forward' with the straight through estimator as its gradient."""
+
+  class StraightThroughEstimator(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, input):
+      return forward(input)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+      return grad_output
+
+  return StraightThroughEstimator.apply
+
+
+def integerize_to_signed_range(x: Tensor, bits: int) -> Tensor:
+  """Round x and clamp it the signed int range for the given number of bits.
+
+  Args:
+    x: The input tensor. Output of scale_max_to_signed_range when quantizing.
+    bits: The number of bits to quantize x to.
+
+  Returns:
+    x rounded and clamped to the appropriate range.
+  """
+  x = ste(torch.round)(x)
+  max_int = get_max_signed_int(bits)
+  return torch.clamp(x, -max_int, max_int)
+
+
+def integerize_to_unsigned_range(x: Tensor, bits: int) -> Tensor:
+  """Floor x and clamp it the unsigned int range for the given number of bits.
+
+  Args:
+    x: The input tensor. Output of scale_max_to_unsigned_range when quantizing.
+    bits: The number of bits to quantize x to.
+
+  Returns:
+    x floored and clamped to the appropriate range.
+  """
+  x = ste(torch.round)(x)
+  return torch.clamp(x, 0, get_max_unsigned_int(bits))
+
+
+def quantize(
+    x: Tensor,  #
+    bits: Optional[int],
+    contraction_dims: List[int]) -> Tuple[Tensor, Optional[Tensor]]:
+  """Quantize x dynamically, returning the quantized value and scale.
+
+  The output should be multiplied by the returned scale after the subsequent
+  linear operation (e.g. matmul, einsum, conv, ...).
+
+  x, x_scale = quantize(x, bits, dims=[1])
+  w, w_scale = quantize(w, bits, dims=[0])
+  y = matmul(x, w)
+  y = y * x_scale * w_scale
+
+  The scale returned has the same shape as x but with 1 in place of each
+  contraction dim. Depending on the op the scale may need to be reshaped before
+  multiplying it with the output.
+
+  Args:
+    x: The tensor to quantize.
+    bits: The number of bits to quantize x to..
+    contraction_dims: The dimensions along which to quantize (complement of the
+      feature dims).
+
+  Returns:
+    Quantized value and scale; the scale has the same rank as x, as we retain
+    reduced dimensions as 1. For the bits=None case, we return x without
+    quantizing and a ones tensor for the scale.
+  """
+  if bits is None:
+    # Normalize dims to all be positive.
+    contraction_dims = [
+        d if d >= 0 else len(x.shape) + d for d in contraction_dims
+    ]
+    # Replace all contraction dims with one.
+    scale_shape = [
+        1 if i in contraction_dims else dim for i, dim in enumerate(x.shape)
+    ]
+    return x, torch.ones(scale_shape, dtype=x.dtype)
+  if (x < 0).any():
```

Comment:
This makes adding quantization to a model quite a bit easier because it removes the need to modify the forward prop code. It adds a ~25% overhead to the runtime of `quantize` however. WDYT?

Reply:
> bit easier because it removes the need to modify the forward prop code.
I probably am missing something here. However, we could always leave comment here on performance and come back to it later to optimize.